lib/pure/ioselectors.nim

```diff
@@ -0,0 +1,1768 @@
+                             "Descriptor is not registered in queue")
+
+    template pollRemove[T](s: Selector[T], sock: cint) =
+      withPollLock(s):
+        var i = 0
+        while i < s.pollcnt:
+          if s.pollfds[i].fd == sock:
+            if i == s.pollcnt - 1:
+              s.pollfds[i].fd = 0
+              s.pollfds[i].events = 0
+              s.pollfds[i].revents = 0
+            else:
+              while i < (s.pollcnt - 1):
+                s.pollfds[i].fd = s.pollfds[i + 1].fd
+                s.pollfds[i].events = s.pollfds[i + 1].events
+                inc(i)
+            break
+          inc(i)
+        dec(s.pollcnt)
+        dec(s.count)
+
+    proc registerHandle*[T](s: Selector[T], fd: SocketHandle,
+                            events: set[Event], data: T) =
+      var fdi = int(fd)
+      s.checkMaxFd(fdi)
+      doAssert(s.fds[fdi].ident == 0)
+      setKey(s, fdi, fdi, {Event.flagHandle} + events, 0, data)
+      s.pollAdd(fdi.cint, events)
+
+    proc updateHandle*[T](s: Selector[T], fd: SocketHandle,
+                          events: set[Event]) =
+      var fdi = int(fd)
+      s.checkMaxFd(fdi)
+      var oe = s.fds[fdi].flags
+      doAssert(s.fds[fdi].ident != 0)
+      doAssert(Event.flagHandle in oe)
+      var ne = events + {Event.flagHandle}
+      if ne != oe:
+        if events != {}:
+          s.pollUpdate(fd.cint, events)
+        else:
+          s.pollRemove(fd.cint)
+        s.fds[fdi].flags = ne
+
+    proc registerTimer*[T](s: Selector[T], timeout: int, oneshot: bool,
+                           data: T): int {.discardable.} =
+      raise newException(ValueError, "Not implemented")
+
+    proc registerSignal*[T](s: Selector[T], signal: int,
+                            data: T): int {.discardable.} =
+      raise newException(ValueError, "Not implemented")
+
+    proc registerProcess*[T](s: Selector[T], pid: int,
+                             data: T): int {.discardable.} =
+      raise newException(ValueError, "Not implemented")
+
+    proc registerEvent*[T](s: Selector[T], ev: SelectEvent, data: T) =
+      var fdi = int(ev.rfd)
+      doAssert(s.fds[fdi].ident == 0)
+      var events = {Event.flagUser, Event.Read}
+      setKey(s, fdi, fdi, events, 0, data)
+      s.pollAdd(fdi.cint, events)
+
+    proc flush*[T](s: Selector[T]) = discard
+
+    template isEmpty*[T](s: Selector[T]): bool =
+      (s.count == 0)
+
+    proc unregister*[T](s: Selector[T], fd: int|SocketHandle|cint) =
+      var fdi = int(fd)
+      if fdi.uint < s.maxFD:
+        if s.fds[fdi].ident != 0 and s.fds[fdi].flags != {}:
+          clearKey(s, fdi)
+          s.pollRemove(fdi.cint)
+
+    proc unregister*[T](s: Selector[T], ev: SelectEvent) =
+      var fdi = int(ev.rfd)
+      if fdi.uint < s.maxFD:
+        if s.fds[fdi].ident != 0 and (Event.flagUser in s.fds[fdi].flags):
+          clearKey(s, fdi)
+          s.pollRemove(fdi.cint)
+
+    proc newEvent*(): SelectEvent =
```

Comment:
I have made it ptr to be allocated in shared memory, so threads can share one object in memory. If one of the threads closes SelectEvent (for some reason), then it will be unusable at all for all threads...

But if SelectEvent becomes an object, then every thread will have copy of SelectEvent and appropriate file descriptors. If one of threads closes SelectEvent, it closes file descriptors too. This can cause EBADF error for other threads on this file descriptors...